### DISCUSSION TITLE
could the performance of fast winding number precomputation be improved?

### DISCUSSION THREAD

**Discussion by wtholliday** [libigl/libigl#1856] [Ideas]
I noticed that the fast winding number precomputation for clouds iterates over all points under the current octree node.

Here: https://github.com/libigl/libigl/blob/96f5386b9a3ff8ad1c2f89b7b77a7db85d86b35b/include/igl/fast_winding_number.cpp#L64

Again here for the higher order approximation terms: https://github.com/libigl/libigl/blob/96f5386b9a3ff8ad1c2f89b7b77a7db85d86b35b/include/igl/fast_winding_number.cpp#L79

It seems like this information could be computed bottom up in the tree.

The first order coefficient is just the area weighted sum of normals, which can be summed up the tree. The second order coefficient seems like it can be summed up the tree as well:

<img width="335" alt="image" src="https://user-images.githubusercontent.com/841144/126086496-4a3c64f0-ebcb-45ef-96f9-86cbde6f7a0a.png">

Does that look correct?

I haven't looked at the 3rd order term yet.

**Comment by alecjacobson** [comment]
Yes, I believe you're right.

**Reply by wtholliday** [reply]
I might attempt a PR, but I'm rather inept with Eigen.

**Comment by wtholliday** [comment]
@alecjacobson Here's what I've got so far: https://github.com/libigl/libigl/compare/main...wtholliday:fast-winding-number-optimization
